File: app/src/main/java/de/christinecoenen/code/zapp/app/player/VideoInfo.java
package de.christinecoenen.code.zapp.app.player;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Objects;

import de.christinecoenen.code.zapp.app.mediathek.model.MediathekShow;
import de.christinecoenen.code.zapp.app.mediathek.model.PersistedMediathekShow;
import de.christinecoenen.code.zapp.app.mediathek.model.Quality;
import de.christinecoenen.code.zapp.app.settings.repository.StreamQualityBucket;
import de.christinecoenen.code.zapp.model.ChannelModel;

public class VideoInfo {

	public static VideoInfo fromShow(PersistedMediathekShow persistedShow) {
		MediathekShow show = persistedShow.getMediathekShow();

		VideoInfo videoInfo = new VideoInfo();
		videoInfo.id = persistedShow.getId();
		videoInfo.url = show.getVideoUrl(Quality.Medium);
		videoInfo.urlHighestQuality = show.getVideoUrl(Quality.High);
		videoInfo.urlLowestQuality = show.getVideoUrl(Quality.Low);
		videoInfo.filePath = persistedShow.getDownloadedVideoPath();
		videoInfo.title = show.getTitle();
		videoInfo.subtitle = show.getTopic();
		videoInfo.subtitleUrl = show.getSubtitleUrl();
		videoInfo.hasDuration = true;

		return videoInfo;
	}

	public static VideoInfo fromChannel(ChannelModel channel) {
		VideoInfo videoInfo = new VideoInfo();
		videoInfo.url = channel.getStreamUrl();
		videoInfo.title = channel.getName();
		videoInfo.subtitle = channel.getSubtitle();
		videoInfo.hasDuration = false;
		return videoInfo;
	}

	private int id = 0;

	@NonNull
	private String url = "";

	private String urlLowestQuality = null;

	private String urlHighestQuality = null;

	@NonNull
	private String title = "";

	private String subtitle;

	@Nullable
	private String subtitleUrl;

	private String filePath;

	private boolean hasDuration = false;

	public int getId() {
		return id;
	}

	@NonNull
	String getPlaybackUrlOrFilePath(StreamQualityBucket quality) {
		if (isOfflineVideo()) {
			return filePath;
		}

		switch (quality) {
			case MEDIUM:
				return url;
			case HIGHEST:
				return urlHighestQuality == null ? url : urlHighestQuality;
			default:
				return urlLowestQuality == null ? url : urlLowestQuality;
		}
	}

	@NonNull
	public String getTitle() {
		return title;
	}

	@NonNull
	public String getSubtitle() {
		return subtitle;
	}

	@Nullable
	String getSubtitleUrl() {
		return subtitleUrl;
	}

	public boolean hasSubtitles() {
		// no subtitle support for downloaded videos yet
		return subtitleUrl != null && !isOfflineVideo();
	}

	public boolean isOfflineVideo() {
		return filePath != null;
	}

	boolean hasDuration() {
		return hasDuration;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		VideoInfo videoInfo = (VideoInfo) o;
		return hasDuration == videoInfo.hasDuration &&
			url.equals(videoInfo.url) &&
			Objects.equals(urlLowestQuality, videoInfo.urlLowestQuality) &&
			Objects.equals(urlHighestQuality, videoInfo.urlHighestQuality) &&
			title.equals(videoInfo.title) &&
			Objects.equals(subtitle, videoInfo.subtitle) &&
			Objects.equals(filePath, videoInfo.filePath) &&
			Objects.equals(subtitleUrl, videoInfo.subtitleUrl);
	}

	@Override
	public int hashCode() {
		return Objects.hash(url, urlLowestQuality, urlHighestQuality, title, subtitle, subtitleUrl, filePath, hasDuration);
	}

	@NonNull
	@Override
	public String toString() {
		return "VideoInfo{" +
			"url='" + url + '\'' +
			", urlLowestQuality='" + urlLowestQuality + '\'' +
			", urlHighestQuality='" + urlHighestQuality + '\'' +
			", title='" + title + '\'' +
			", subtitle='" + subtitle + '\'' +
			", subtitleUrl='" + subtitleUrl + '\'' +
			", filePath='" + filePath + '\'' +
			", hasDuration=" + hasDuration +
			'}';
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/player/VideoInfo.kt
package de.christinecoenen.code.zapp.app.player

import android.text.TextUtils
import de.christinecoenen.code.zapp.models.shows.PersistedMediathekShow
import de.christinecoenen.code.zapp.models.shows.Quality
import de.christinecoenen.code.zapp.app.settings.repository.StreamQualityBucket
import de.christinecoenen.code.zapp.models.channels.ChannelModel

data class VideoInfo(
	var id: Int = 0,
	var title: String,
	var url: String,
	var urlLowestQuality: String? = null,
	var urlHighestQuality: String? = null,
	var subtitle: String? = null,
	var subtitleUrl: String? = null,
	var filePath: String? = null,
	var hasDuration: Boolean = false
) {

	companion object {

		@JvmStatic
		fun fromShow(persistedShow: PersistedMediathekShow): VideoInfo {
			val show = persistedShow.mediathekShow

			return VideoInfo(
				title = show.title,
				url = show.videoUrl
			).apply {
				id = persistedShow.id
				urlHighestQuality = show.getVideoUrl(Quality.High)
				urlLowestQuality = show.getVideoUrl(Quality.Low)
				filePath = persistedShow.downloadedVideoPath
				subtitle = show.topic
				subtitleUrl = show.subtitleUrl
				hasDuration = true
			}
		}

		@JvmStatic
		fun fromChannel(channel: ChannelModel): VideoInfo {
			return VideoInfo(
				title = channel.name,
				url = channel.streamUrl
			).apply {
				url = channel.streamUrl
				subtitle = channel.subtitle
				hasDuration = false
			}
		}

	}

	val hasSubtitles: Boolean
		// no subtitle support for downloaded videos yet
		get() = !TextUtils.isEmpty(subtitleUrl) && !isOfflineVideo

	val isOfflineVideo: Boolean
		get() = filePath != null

	fun getPlaybackUrlOrFilePath(quality: StreamQualityBucket): String {
		return if (isOfflineVideo) {
			filePath!!
		} else when (quality) {
			StreamQualityBucket.MEDIUM -> url
			StreamQualityBucket.HIGHEST -> urlHighestQuality ?: url
			else -> urlLowestQuality ?: url
		}
	}
}
